kiwmi/desktop/view.c

```diff
@@ -259,3 +284,191 @@ view_create(
+    wl_list_remove(&child->surface_destroy.link);
+    wl_list_remove(&child->owner_destroy.link);
+
+    free(child);
+}
+
+static void
+view_child_subsurface_owner_destroy_notify(
+    struct wl_listener *listener,
+    void *UNUSED(data))
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, owner_destroy);
+    view_child_destroy(child);
+}
+
+static void
+view_child_surface_destroy_notify(
+    struct wl_listener *listener,
+    void *UNUSED(data))
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, surface_destroy);
+    view_child_destroy(child);
+}
+
+static void
+view_child_commit_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child = wl_container_of(listener, child, commit);
+    view_child_damage(child);
+}
+
+static void
+view_child_new_subsurface_notify(struct wl_listener *listener, void *data)
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, new_subsurface);
+    struct wlr_subsurface *subsurface = data;
+    view_child_subsurface_create(child, child->view, subsurface);
+}
+
+static void
+view_child_map_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child = wl_container_of(listener, child, map);
+    child->mapped                  = true;
+    view_child_damage(child);
+}
+
+static void
+view_child_unmap_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child = wl_container_of(listener, child, unmap);
+    child->mapped                  = false;
+    view_child_damage(child);
+}
+
+static void
+view_child_view_unmap_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, view_unmap);
+    child->mapped = false;
+    view_child_damage(child);
+}
+
+struct kiwmi_view_child *
+view_child_create(
+    struct kiwmi_view *view,
+    struct wlr_surface *wlr_surface,
+    enum kiwmi_view_child_type type,
+    const struct kiwmi_view_child_impl *impl)
+{
+    struct kiwmi_view_child *child = calloc(1, sizeof(*child));
+    if (!child) {
+        wlr_log(WLR_ERROR, "Failed to allocate view_child");
+        return NULL;
+    }
+
+    child->type        = type;
+    child->impl        = impl;
+    child->view        = view;
+    child->wlr_surface = wlr_surface;
+
+    wl_list_init(&child->link);
+    wl_list_init(&child->children);
+    child->mapped = false;
+
+    child->commit.notify = view_child_commit_notify;
+    wl_signal_add(&wlr_surface->events.commit, &child->commit);
+
+    child->surface_destroy.notify = view_child_surface_destroy_notify;
+    wl_signal_add(&wlr_surface->events.destroy, &child->surface_destroy);
+
+    child->new_subsurface.notify = view_child_new_subsurface_notify;
+    wl_signal_add(&wlr_surface->events.new_subsurface, &child->new_subsurface);
+
+    child->view_unmap.notify = view_child_view_unmap_notify;
```

Comment:
I was copying this from Sway. It only listens to `unmap` on the view. But looking at https://github.com/swaywm/sway/commit/017a7c4da169ace601f282c87a9aaabc138a34d3, there seems to be no reason not to handle `map` as well.

Reply:
Ok epic. I'd say don't handle if we don't need to.

Reply:
I wonder if it’s just that views never unmap other than before destroying or if it’s completely correct.

(Marking as unresolved only so that you see this, feel free to re-resolve.)

Reply:
I haven’t found where these (or rather, the code that calls them) are called, but wlroots seem to handle parents for us (so the only thing we have to care about is the invalid pointer).

https://github.com/swaywm/wlroots/blob/46c42e55c6e0b08d9e7db989d6a10f073525e999/types/wlr_surface.c#L1026..L1080